### PR TITLE
Rdm 5762 use globus search backend

### PR DESCRIPTION
**Pull request by adammckenzie**
Hi Paul,

If you have time while I am gone to review this that would be awesome. It is currently deployed on dev1 and is the same contents as the akm220 branch if you want to check it out on dev2. 

This changes the backend from using Solr to using Globus Search. Searching, Map, all of the facets and sorting should be functioning correctly. There are calendar pickers for choosing publication date for the dataset. You can search for source repos and authors using the search box above the list. Choosing a filter from a facet moves it to the top and shows it even when there are zero results. All of the facets now show even if there is nothing to show. 

Limitation: Globus Search backend currently doesn't show results past 9999 or page 500 so paging to the end will show zero results. 

Future improvements:
The following files are currently copied verbatim from blacklight as I couldn't autoload them without causing other errors from their source usage (both modules wanted to own them causing an error in the other one). When I get back I will either figure out how to autoload them properly or just create new files that extend the originals keeping all of their functionality. 
lib/blacklight/globus_search/response/facets.rb
lib/blacklight/globus_search/response/more_like_this.rb
lib/blacklight/globus_search/response/pagination_methods.rb
lib/blacklight/globus_search/response/params.rb
lib/blacklight/globus_search/response/response.rb
lib/blacklight/globus_search/response/spelling.rb

I created the issue: 
https://computecanada.atlassian.net/browse/RDM-6175

To fix the issue.

### CODE REVIEW

File: lib/blacklight/globus_search/response/facets.rb [hunk 1]
```diff
@@ -0,0 +1,310 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/facets.rb:1]: Copied from blacklight

File: lib/blacklight/globus_search/response/more_like_this.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/more_like_this.rb:1]: Copied from blacklight

File: lib/blacklight/globus_search/response/pagination_methods.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/pagination_methods.rb:1]: Copied from blacklight

File: lib/blacklight/globus_search/response/params.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/params.rb:1]: Copied from blacklight

File: lib/blacklight/globus_search/response/response.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/response.rb:1]: Copied from blacklight

File: lib/blacklight/globus_search/response/spelling.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
```
Comment (adammckenzie) [inline lib/blacklight/globus_search/response/spelling.rb:1]: Copied from blacklight

**Comment by pdante-ubc** [comment]
Went to Dev1 and all of the facets are giving me an error when I click a checkbox if I haven't done some sort of text search first. For example, if I just click "Search here" on the landing page and then directly try to click on any facet checkbox it gives me a something went wrong error. From the logs it looks like it is passing nil for the text search parameter, and that is breaking things. Once a search has been done using the text search, if if that search had nothing in the text box, the url parameter for q is now always there even if it is just `q=`

**Comment by pdante-ubc** [comment]
When I try and search in the Author facet, it mostly works but there is still some weird stuff going on. When I search for "City of" in that search box, it does bring up a bunch of authors that include "City of..." but it also has some other results such as "Aukes, Pieter" or "Government of New Brunswick". For the latter, maybe it's matching on the "of", but what is it matching on the former?

**Comment by pdante-ubc** [comment]
Per: "Limitation: Globus Search backend currently doesn't show results past 9999 or page 500 so paging to the end will show zero results."

If we're going to leave it like that for now, maybe we should limit the jump to page # option at the bottom to max out at page 500 and have some sort of text indicating that the searcher should filter more to be able to see all the results (it's a semi-counterintuitive statement that I'm not sure how to word)?

**Comment by pdante-ubc** [comment]
Should we add the option to exclude a facet result (e.g., I want to select all repos except "Open Data Canada")? Not sure if this is a useful thing to do .

**Comment by pdante-ubc** [comment]
The facets only show the first 7(?) options for that facet, can we do something to allow people to see more? Relatedly, can we allow people to sort the facet options by more than just number of records, such as alphabetically (forward or backward) or by number of records small to large?

**Comment by pdante-ubc** [comment]
Other than than my comments, it seems like the switch to Globus facets is working correctly.

**Comment by pdante-ubc** [comment]
Seems like there might be an existing suggestions when entering a search for author/repo/title, but is is currently not being used (lib/blacklight/globus_search/repository.rb, line 30). I'm assuming this is currently not setup for working with Globus as I think it is the default code. It could be nice to have suggestions popping up after the first 3-5 characters are typed.

**Review by pdante-ubc — COMMENTED** [review]
One bug and a couple suggestions, but otherwise looks good.

**Comment by adammckenzie** [comment]
> Per: "Limitation: Globus Search backend currently doesn't show results past 9999 or page 500 so paging to the end will show zero results."
> 
> If we're going to leave it like that for now, maybe we should limit the jump to page # option at the bottom to max out at page 500 and have some sort of text indicating that the searcher should filter more to be able to see all the results (it's a semi-counterintuitive statement that I'm not sure how to word)?

Yeah we will discuss in the meeting with Globus when we show what we have so far. Depending on whether they increase the limit will dictate whether we need to change the UI.

**Comment by adammckenzie** [comment]
> Should we add the option to exclude a facet result (e.g., I want to select all repos except "Open Data Canada")? Not sure if this is a useful thing to do .

That is a big enough of a feature we would probably need to discuss it in the redesign meeting.

**Comment by adammckenzie** [comment]
> The facets only show the first 7(?) options for that facet, can we do something to allow people to see more? Relatedly, can we allow people to sort the facet options by more than just number of records, such as alphabetically (forward or backward) or by number of records small to large?

The more we show the bigger the page grows so there is a tricky trade off here that might require testing. That is why we have the search for people to discover other options. They will mostly use the search box anyway. 

Sorting is out of scope.

**Comment by adammckenzie** [comment]
> Seems like there might be an existing suggestions when entering a search for author/repo/title, but is is currently not being used (lib/blacklight/globus_search/repository.rb, line 30). I'm assuming this is currently not setup for working with Globus as I think it is the default code. It could be nice to have suggestions popping up after the first 3-5 characters are typed.

We need to talk about this in the redesign meeting. The suggestions I think are less helpful than we want but we should take a look at them for sure. For right now we can remove some comments to show them again. There would be some UI work that would also need to be done to fix them, they appear behind certain UI elements due to z settings.

**Comment by adammckenzie** [comment]
> Went to Dev1 and all of the facets are giving me an error when I click a checkbox if I haven't done some sort of text search first. For example, if I just click "Search here" on the landing page and then directly try to click on any facet checkbox it gives me a something went wrong error. From the logs it looks like it is passing nil for the text search parameter, and that is breaking things. Once a search has been done using the text search, if if that search had nothing in the text box, the url parameter for q is now always there even if it is just `q=`

Hi Paul,

Did you happen to test this after making your changes in "e8454a73cd324e1f9326b1e263a489c31df92245". There was a couple of syntax errors in you change that was causing issues so I have reverted it. I can't reproduce this now with either a bbox search and no query, an empty query, or a single space in the query. There might still be a bug somewhere but I am not sure exactly which sequence would trigger it.

I think for pull requests the more typical way to suggest changes is to add a comment rather than pushing changes directly to the branch as it took me some time to figure out what exactly had gone wrong until I had checked the git log.

**Comment by pdante-ubc** [comment]
> > Went to Dev1 and all of the facets are giving me an error when I click a checkbox if I haven't done some sort of text search first. For example, if I just click "Search here" on the landing page and then directly try to click on any facet checkbox it gives me a something went wrong error. From the logs it looks like it is passing nil for the text search parameter, and that is breaking things. Once a search has been done using the text search, if if that search had nothing in the text box, the url parameter for q is now always there even if it is just `q=`
> 
> Hi Paul,
> 
> Did you happen to test this after making your changes in "e8454a73cd324e1f9326b1e263a489c31df92245". There was a couple of syntax errors in you change that was causing issues so I have reverted it. I can't reproduce this now with either a bbox search and no query, an empty query, or a single space in the query. There might still be a bug somewhere but I am not sure exactly which sequence would trigger it.
> 
> I think for pull requests the more typical way to suggest changes is to add a comment rather than pushing changes directly to the branch as it took me some time to figure out what exactly had gone wrong until I had checked the git log.

Sorry about the main git push. Everything was working when I pulled main into cluster, but then broken when I merged the merged cluster/main back into cluster. It turned out to be an issue with the versions of Gem files. I will make sure to make those small changes via a pull request in the future.

As for the bug, it doesn't seem to be happening on dev1 anymore, so I guess it somehow sorted itself out.

**Comment by adammckenzie** [comment]
> > > Went to Dev1 and all of the facets are giving me an error when I click a checkbox if I haven't done some sort of text search first. For example, if I just click "Search here" on the landing page and then directly try to click on any facet checkbox it gives me a something went wrong error. From the logs it looks like it is passing nil for the text search parameter, and that is breaking things. Once a search has been done using the text search, if if that search had nothing in the text box, the url parameter for q is now always there even if it is just `q=`
> > 
> > 
> > Hi Paul,
> > Did you happen to test this after making your changes in "e8454a73cd324e1f9326b1e263a489c31df92245". There was a couple of syntax errors in you change that was causing issues so I have reverted it. I can't reproduce this now with either a bbox search and no query, an empty query, or a single space in the query. There might still be a bug somewhere but I am not sure exactly which sequence would trigger it.
> > I think for pull requests the more typical way to suggest changes is to add a comment rather than pushing changes directly to the branch as it took me some time to figure out what exactly had gone wrong until I had checked the git log.
> 
> Sorry about the main git push. Everything was working when I pulled main into cluster, but then broken when I merged the merged cluster/main back into cluster. It turned out to be an issue with the versions of Gem files. I will make sure to make those small changes via a pull request in the future.
> 
> As for the bug, it doesn't seem to be happening on dev1 anymore, so I guess it somehow sorted itself out.

That's ok these things happen. Thanks for reviewing all of this!

**Pull request merged**